File: kotlin/src/main/kotlin/tech/sia/indexdsdk/Wrappers.kt
/**
 * Idiomatic Kotlin wrappers for the Sia SDK.
 *
 * This module provides convenience classes and extension functions that make the SDK
 * feel native to Kotlin developers, wrapping the low-level Reader/Writer
 * traits with standard Java/Kotlin I/O support.
 */
package tech.sia.indexdsdk

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.OutputStream

/**
 * Adapts an [InputStream] to the [Reader] trait.
 *
 * This allows you to upload data from any source that supports the standard
 * Java/Kotlin InputStream interface (files, ByteArrayInputStream, etc.).
 *
 * Example:
 * ```kotlin
 * // From bytes
 * val reader = StreamReader(ByteArrayInputStream(data))
 *
 * // From file
 * val reader = StreamReader(FileInputStream("data.bin"))
 * ```
 */
class StreamReader(
    private val stream: InputStream,
    private val chunkSize: Int = 65536,
) : Reader {
    override suspend fun read(): ByteArray {
        val buffer = ByteArray(chunkSize)
        val bytesRead = stream.read(buffer)
        return if (bytesRead <= 0) ByteArray(0) else buffer.copyOf(bytesRead)
    }
}

/**
 * Adapts an [OutputStream] to the [Writer] trait.
 *
 * This allows you to download data to any destination that supports the
 * standard Java/Kotlin OutputStream interface (files, ByteArrayOutputStream, etc.).
 *
 * Example:
 * ```kotlin
 * // To ByteArrayOutputStream
 * val buffer = ByteArrayOutputStream()
 * val writer = StreamWriter(buffer)
 *
 * // To file
 * val writer = StreamWriter(FileOutputStream("output.bin"))
 * ```
 */
class StreamWriter(
    private val stream: OutputStream,
) : Writer {
    override suspend fun write(data: ByteArray) {
        if (data.isNotEmpty()) {
            stream.write(data)
        }
    }
}

/**
 * Upload bytes directly to the Sia network.
 *
 * Example:
 * ```kotlin
 * val obj = sdk.uploadBytes("hello, world!".toByteArray())
 * ```
 */
suspend fun Sdk.uploadBytes(
    data: ByteArray,
    options: UploadOptions = UploadOptions(),
): PinnedObject = upload(StreamReader(ByteArrayInputStream(data)), options)

/**
 * Download an object and return its contents as bytes.
 *
 * Example:
 * ```kotlin
 * val data = sdk.downloadBytes(obj)
 * ```
 */
suspend fun Sdk.downloadBytes(
    obj: PinnedObject,
    options: DownloadOptions = DownloadOptions(),
): ByteArray {
    val buffer = ByteArrayOutputStream()
    download(StreamWriter(buffer), obj, options)
    return buffer.toByteArray()
}

File: python/sia_indexd/wrappers.py
"""Idiomatic Python wrappers for the Sia SDK.

This module provides convenience classes and functions that make the SDK
feel native to Python developers, wrapping the low-level Reader/Writer
traits with standard file-like object support.
"""

import asyncio
from io import BytesIO
from typing import BinaryIO, Optional

from .sia_indexd.indexd_ffi import (
    AppMeta,
    Builder as _Builder,
    DownloadOptions,
    PinnedObject,
    Reader,
    Sdk,
    UploadOptions,
    Writer,
    uniffi_set_event_loop,
)


class Builder:
    """Wrapper around the UniFFI Builder that auto-initializes the event loop."""

    def __init__(self, base_url: str):
        try:
            uniffi_set_event_loop(asyncio.get_running_loop())
        except RuntimeError:
            pass  # No running loop yet
        self._inner = _Builder(base_url)

    async def request_connection(self, app_meta: AppMeta) -> None:
        return await self._inner.request_connection(app_meta)

    def response_url(self) -> str:
        return self._inner.response_url()

    async def wait_for_approval(self) -> None:
        return await self._inner.wait_for_approval()

    async def register(self, mnemonic: str) -> Sdk:
        return await self._inner.register(mnemonic)


class BytesReader(Reader):
    """Adapts any file-like object to the Reader trait.

    This allows you to upload data from any source that supports the
    standard Python binary read interface (files, BytesIO, etc.).

    Example:
        # From bytes
        reader = BytesReader(BytesIO(b"hello"))

        # From file
        with open("data.bin", "rb") as f:
            reader = BytesReader(f)
    """

    def __init__(self, stream: BinaryIO, chunk_size: int = 65536):
        """Create a BytesReader from a file-like object.

        Args:
            stream: Any object with a read(n) method returning bytes.
            chunk_size: Size of chunks to read (default 64KiB).
        """
        self._stream = stream
        self._chunk_size = chunk_size

    async def read(self) -> bytes:
        """Read the next chunk of data."""
        return self._stream.read(self._chunk_size)


class BytesWriter(Writer):
    """Adapts any file-like object to the Writer trait.

    This allows you to download data to any destination that supports
    the standard Python binary write interface (files, BytesIO, etc.).

    Example:
        # To BytesIO
        buf = BytesIO()
        writer = BytesWriter(buf)

        # To file
        with open("output.bin", "wb") as f:
            writer = BytesWriter(f)
    """

    def __init__(self, stream: BinaryIO):
        """Create a BytesWriter from a file-like object.

        Args:
            stream: Any object with a write(data) method.
        """
        self._stream = stream

    async def write(self, data: bytes) -> None:
        """Write a chunk of data."""
        if len(data) > 0:
            self._stream.write(data)


async def upload_bytes(
    sdk: Sdk,
    data: bytes,
    options: Optional[UploadOptions] = None,
) -> PinnedObject:
    """Upload bytes directly to the Sia network.

    Args:
        sdk: The SDK instance.
        data: The bytes to upload.
        options: Optional upload options.

    Returns:
        The pinned object representing the uploaded data.

    Example:
        obj = await upload_bytes(sdk, b"hello, world!")
    """
    return await sdk.upload(BytesReader(BytesIO(data)), options or UploadOptions())


async def download_bytes(
    sdk: Sdk,
    obj: PinnedObject,
    options: Optional[DownloadOptions] = None,
) -> bytes:
    """Download an object and return its contents as bytes.

    Args:
        sdk: The SDK instance.
        obj: The pinned object to download.
        options: Optional download options.

    Returns:
        The downloaded data as bytes.

    Example:
        data = await download_bytes(sdk, obj)
    """
    buf = BytesIO()
    await sdk.download(BytesWriter(buf), obj, options or DownloadOptions())
    return buf.getvalue()
